### PR TITLE
adding test flags to control load size and test rook resilience 

### LINKED ISSUE #1216: Implement Performance Test for Block and object Storage.
Performance test for Block Storage would include tweaking the long haul test and converting it into a load test and Collecting metrics.  The plan is to use Prometheus to gather node, pod and ceph metrics and then user Grafana to vizualize them.  We Also need to implement a solution to take periodic snapshtos of Grafana and archive them along with test logs.

### PR DESCRIPTION
**Pull request by dangula**
Description of your changes: 
- added new tests flag `--load_size` to control amount of data being written -S,M or L(less than 1M,1M and 10M respectively)
 - added new test flag `--enable_chaos` - if this flag is present, random pods in rook cluster will be killed peridiocally
 - update longhaul test pipeline (added new flags)
 - updated docs
fix #1216

### CODE REVIEW

**Comment by travisn** [comment]
why is longhaul_util.go over 12K lines? i'm having trouble viewing it

**Comment by jbw976** [comment]
@travisn looks like it's from `func getGrafanaConfigMaps`, which I don't think is entirely needed.  I don't think GUI grafana dashboards are needed for the longhaul tests.

**Comment by dangula** [comment]
The Grafana ConfigMap came down to 12K after i stipped  out all the things that were not needed(following examples from  promethoes and grafana side). Currently, promethous and Grananfa are monitoring node and pod metrics (memory,cpu, disk io etc.. at pod and nod level). 
  
   After running a few perf tests on pulbic cloud and baremetal environments,  I too feel like Prometheus and grafana dashboards should probably not be part of test framework. Mainly because  public clould feels like not a good fit for perf test( as they have their own iops limits). Perf test makes sense in baremetal env where we can push rook and measure the throughput.  I can remove the prometheus and grafana (or just grafana ) from the PR. Create a new rook-utils repo  or new  folder under /cluster/exampels and put all  the prometheus and grafana specs in there (along with instructions), so that users can choose to start Grafana to minitor their rook cluster.

**Comment by travisn** [comment]
@dangula what if we just added a help topic that provided a pointer to grafana and prometheus documentation to get those started? I would think all the examples could be found in the other repos

**Comment by dangula** [comment]
That could work too, There are good examples online on getting prometheus and grafana up to monitor a k8s cluster. I will update the PR shortly.

File: tests/README.md [hunk 1]
```diff
@@ -80,16 +80,15 @@ In addition to standard go tests parameters, the following custom parameters are
...
 rook_image | rook image name to be installed | valid image name | rook/rook
 toolbox_image | toolbox image name to be installed | valid image name | rook/toolbox
 skip_install_rook | skips installing rook (if already installed) | true or false  | false
-load_parallel_runs | performs concurrent operations (optional used for load test) | any number | 20
-
-If the `install_rook` flag is set to false, then all the other flags are ignored
-and tests are run without rook being installed and setup. Use this flag to run tests against
-a pre-installed/configured rook.
-
+load_parallel_runs | performs concurrent operations (optional used for load test only) | any number | 20
```
Comment (travisn) [inline tests/README.md]: suggestion: in parenthesis keep it brief with: `(load test only)`. The same thing applies for all these load test settings

File: tests/README.md [hunk 2]
```diff
@@ -80,16 +80,15 @@ In addition to standard go tests parameters, the following custom parameters are
...
+load_volumes | number of volumes ( block,object store or shared file systems) to be created during load test (optional used for load test only)| >1 | 1
+load_time | number of seconds load test run for (optional used for load test only) | >1 | 1800
+load_size | load profile - choice S,M or L(3M,10M and 50M per thread - optional used for load test only) | S,M or L | M
```
Comment (travisn) [inline tests/README.md]: what about this description? 
`load_size | size of profile to load (3M, 10M, or 50M per thread - load test only) | S, M, or L | M`

File: tests/README.md [hunk 3]
```diff
@@ -111,23 +110,24 @@ go test -v -timeout 1800s -run SmokeSuite github.com/rook/rook/tests/integration
...
 If the `skip_install_rook` flag is set to true, then rook is not uninstalled either. 
 
 #### Run Longhaul Tests
-Using the `go test -count n` option, any tests can be repeated `n` number of times to simulate load or longhaul test. Although 
-any test can be converted to a longhaul test, it's ideal to run integration tests under load for an extended period. Also a new custom test flag `load_parallel_runs` is added to control the number of concurrent operations being performed.
-For example look at the [block long haul test](/tests/block/k8s/longhaul/basicBlockonghaul_test.go)
- 
- To run a longhaul test you can run any integration test with `-count` and `--load_parallel_runs` options
- e.g.
+[longhaul](/tests/block/k8s/longhaul) tests are integration tests that run for extended period of time, the load profile can be configured
```
Comment (travisn) [inline tests/README.md]: This paragraph needs complete sentences, capitalization, and punctuation.

File: tests/README.md (hunk 2 above)
Comment (travisn) [reply tests/README.md]: where do those sizes come from? Should you just allow the number to be used directly instead of using S,M,L?

File: tests/README.md [hunk 4]
```diff
@@ -111,23 +110,24 @@ go test -v -timeout 1800s -run SmokeSuite github.com/rook/rook/tests/integration
...
+For example look at the [Object long haul test](/tests/block/k8s/longhaul/object_test.go)
+  e.g.
  ```
- go test -run TestK8sBlockLongHaul github.com/rook/rook/tests/longhaul --load_parallel_runs=20 -count=1000
+ go test -run TestObjectLongHaul github.com/rook/rook/tests/longhaul --load_parallel_runs=20 --load_time 1800 --load_size S --load_volumes 3
  ```
  The Longhaul test just like other test is going to install rook if it's not already installed, but it is not going to clean up test data or uninstall rook after the run. 
  Longhaul test is designed to run multiple times on the same setup and installation of rook to tests its stability. Test Data and rook should be cleaned up manually after the test.
  
+ You can deploy Prometheus and Grafana to monitor rook. Here a couple of helpful links to get prometheus and grafana started and collect metrics on pod and cluster level -
```
Comment (travisn) [inline tests/README.md]: You might add another sentence describing why it will be useful

File: tests/longhaul/base_block_test.go [hunk 5]
```diff
@@ -99,24 +94,35 @@ func performBlockOperations(installer *installer.InstallHelper, db *utils.MySQLH
...
 	wg.Wait()
 }
 
-func dbOperation(db *utils.MySQLHelper, wg *sync.WaitGroup, runtime int) {
+func dbOperation(db *utils.MySQLHelper, wg *sync.WaitGroup, runtime int, loadSize string) {
 	defer wg.Done()
+	ds := 100000
+	switch strings.ToUpper(loadSize) {
+	case "S":
+		ds = 105000 //.1M * 5 columns * 6 = 3M per thread
```
Comment (travisn) [inline tests/longhaul/base_block_test.go]: i'd suggest letting the compiler compute these numbers for you.

File: tests/longhaul/base_block_test.go [hunk 6]
```diff
@@ -155,6 +161,13 @@ func (o BaseLoadTestOperations) SetUp() {
 		isRookInstalled, err := o.installer.InstallRookOnK8sWithHostPathAndDevices(o.namespace, "bluestore", "/temp/rookBackup", false, true, 3)
 		require.NoError(o.T(), err)
 		require.True(o.T(), isRookInstalled)
+
+	}
+
+	// Enable chaos monkey if enable_chaos flag is present
+	if o.installer.Env.EnableChaos {
+		c := NewChaosHelper(o.namespace, o.kh)
+		go c.Monkey()
```
Comment (travisn) [inline tests/longhaul/base_block_test.go]: how does the chaos monkey know when to stop? Does the whole process terminate when the load test is done?

File: tests/longhaul/longhaul_util.go [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+}
+
+func NewChaosHelper(namespace string, kh *utils.K8sHelper) *ChaosHelper {
+	return &ChaosHelper{namespace: namespace, kh: kh}
+
+}
+
+func (h *ChaosHelper) Monkey() {
+	listOpts := metav1.ListOptions{LabelSelector: "app=rook-operator"}
+	pl, err := h.kh.Clientset.CoreV1().Pods(installer.SystemNamespace(h.namespace)).List(listOpts)
```
Comment (travisn) [inline tests/longhaul/longhaul_util.go]: why do you list the pods in the system namespace here?

File: tests/longhaul/longhaul_util.go [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+				logger.Infof("Error getting list of pods ,err ->%v", err)
+				return
+			}
+			kp := pods.Items[rand.Intn(len(pods.Items)-1)]
+			logger.Info("Killing pod with Name= %s", kp.Name)
+			propagation := metav1.DeletePropagationForeground
+			delOptions := &metav1.DeleteOptions{PropagationPolicy: &propagation}
+			err = h.kh.Clientset.CoreV1().Pods(h.namespace).Delete(kp.Name, delOptions)
+			if err != nil {
+				logger.Infof("Cannot delete pod %s, err->%v", kp.Name, err)
```
Comment (travisn) [inline tests/longhaul/longhaul_util.go]: you need a `continue` statement here

File: tests/longhaul/longhaul_util.go [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+	listOpts := metav1.ListOptions{LabelSelector: "app=rook-operator"}
+	pl, err := h.kh.Clientset.CoreV1().Pods(installer.SystemNamespace(h.namespace)).List(listOpts)
+	if err != nil {
+		logger.Infof("Cannot check status of rook-operator pod,err ->%v", err)
+		return
+	}
+	creationTimeStamp := pl.Items[0].CreationTimestamp
+	t := time.Unix(0, creationTimeStamp.UnixNano())
+	for {
+		elapsed := time.Since(t)
```
Comment (travisn) [inline tests/longhaul/longhaul_util.go]: why not put the Sleep at the beginning of the loop? Then you don't need to check at all for elapsed minutes

File: tests/longhaul/base_block_test.go (hunk 5 above)
Comment (dangula) [reply tests/longhaul/base_block_test.go]: i tried that but in  sizeof just kept giving size varriable not size of string or data the varaible is referencing to. So ended up using number of chars instead.

File: tests/longhaul/longhaul_util.go (hunk 9 above)
Comment (dangula) [reply tests/longhaul/longhaul_util.go]: thought about a few more egde cases and putting the sleep defore the loop makes sense and avoids wierd edge cases - eg users runs block load test for 30 min, and then runs object.  Monkey might kill a pod while we are creating a object store causing expected failures.

File: tests/README.md [hunk 10]
```diff
@@ -80,16 +80,15 @@ In addition to standard go tests parameters, the following custom parameters are
...
 toolbox_image | toolbox image name to be installed | valid image name | rook/toolbox
 skip_install_rook | skips installing rook (if already installed) | true or false  | false
-load_parallel_runs | performs concurrent operations (optional used for load test) | any number | 20
-
-If the `install_rook` flag is set to false, then all the other flags are ignored
-and tests are run without rook being installed and setup. Use this flag to run tests against
-a pre-installed/configured rook.
-
+load_parallel_runs | performs concurrent operations (load test only) | any number | 20
+load_volumes | number of volumes ( block,object store or shared file systems) to be created during load test (load test only)| >1 | 1
```
Comment (travisn) [inline tests/README.md]: 1. i'd suggest the following spacing:
`(block, object store, or shared file systems)`
2. do you mount object store volumes? or are you referring only to block and file?
3. you can remove "to be created during load test" since you already said that in parenthesis

File: tests/README.md [hunk 11]
```diff
@@ -80,16 +80,15 @@ In addition to standard go tests parameters, the following custom parameters are
...
+load_time | number of seconds load test run for (load test only) | >1 | 1800
```
Comment (travisn) [inline tests/README.md]: `number of seconds to run (load test only)`

File: tests/README.md [hunk 12]
```diff
@@ -111,24 +110,29 @@ go test -v -timeout 1800s -run SmokeSuite github.com/rook/rook/tests/integration
...
 If the `skip_install_rook` flag is set to true, then rook is not uninstalled either. 
 
 #### Run Longhaul Tests
-Using the `go test -count n` option, any tests can be repeated `n` number of times to simulate load or longhaul test. Although 
-any test can be converted to a longhaul test, it's ideal to run integration tests under load for an extended period. Also a new custom test flag `load_parallel_runs` is added to control the number of concurrent operations being performed.
-For example look at the [block long haul test](/tests/block/k8s/longhaul/basicBlockonghaul_test.go)
- 
- To run a longhaul test you can run any integration test with `-count` and `--load_parallel_runs` options
- e.g.
+[longhaul](/tests/block/k8s/longhaul) Tests are integration tests that run for extended period of time. A load profile can be configured
```
Comment (travisn) [inline tests/README.md]: "longhaul" is the first word of the sentence, not "Tests".

File: tests/README.md [hunk 13]
```diff
@@ -111,24 +110,29 @@ go test -v -timeout 1800s -run SmokeSuite github.com/rook/rook/tests/integration
...
+using the following load test flags 
++ `load_parallel_runs`: controls number of concurrent operations reading and writing data
```
Comment (travisn) [inline tests/README.md]: You list these flags here and in the table above? I would suggest removing them from the table above. Then it will keep the main tests simple, and people running load tests can look here for other flags. And maybe you could create a new table here instead of the bulleted list.

File: tests/README.md [hunk 14]
```diff
@@ -111,24 +110,29 @@ go test -v -timeout 1800s -run SmokeSuite github.com/rook/rook/tests/integration
...
++ `load_time` : controls how the long the test runs for
++ `load_size` : controls size of data and 
++ `load_volumes` : controls the number of concurrent volumes under test(file,block or object)
+  e.g.
  ```
- go test -run TestK8sBlockLongHaul github.com/rook/rook/tests/longhaul --load_parallel_runs=20 -count=1000
+ go test -run TestObjectLongHaul github.com/rook/rook/tests/longhaul --load_parallel_runs=20 --load_time 1800 --load_size S --load_volumes 3
```
Comment (travisn) [inline tests/README.md]: "S" is not very readable. I would suggest spelling out small, medium, large

File: tests/README.md [hunk 15]
```diff
@@ -111,24 +110,29 @@ go test -v -timeout 1800s -run SmokeSuite github.com/rook/rook/tests/integration
...
  ```
  The Longhaul test just like other test is going to install rook if it's not already installed, but it is not going to clean up test data or uninstall rook after the run. 
  Longhaul test is designed to run multiple times on the same setup and installation of rook to tests its stability. Test Data and rook should be cleaned up manually after the test.
  
  
+ You can measure memory,cpu,ipos,throughput etc. on a cluster, node and pod level using Prometheus and visualize it using Grafana. This would be useful in measuring performance and health of the cluster.
```
Comment (travisn) [inline tests/README.md]: `You can measure memory, cpu, IOPS, throughput, and other node and pod metrics using Prometheus. The metrics collected during the tests can be visualized with Grafana.`

File: tests/README.md (hunk 10 above)
Comment (dangula) [reply tests/README.md]: we just mount block and file. I will actually remove file and block, feels redundant.

File: tests/README.md (hunk 11 above)
Comment (dangula) [reply tests/README.md]: fixed.

File: tests/README.md (hunk 12 above)
Comment (dangula) [reply tests/README.md]: fixed.

File: tests/README.md (hunk 13 above)
Comment (dangula) [reply tests/README.md]: done.

File: tests/README.md (hunk 15 above)
Comment (dangula) [reply tests/README.md]: Done.

File: tests/README.md (hunk 14 above)
Comment (dangula) [reply tests/README.md]: done.
